tsl/src/data_node.c

```diff
@@ -513,6 +513,17 @@ data_node_bootstrap_extension(TSConnection *conn)
 								  " WITH SCHEMA %s VERSION %s CASCADE",
 								  schema_name_quoted,
 								  quote_literal_cstr(ts_extension_get_version()));
+
+		/*
+		 * Since PG15, by default, non-superuser accounts are not allowed to
+		 * create tables in public schema of databases they don't own. This
+		 * default can be changed manually. The following query ensures that the
+		 * permissions are going to be the same regardless of the used PostgreSQL
+		 * version.
+		 *
+		 * https://git.postgresql.org/gitweb/?p=postgresql.git;a=commitdiff;h=b073c3cc
+		 */
+		remote_connection_cmdf_ok(conn, "GRANT CREATE ON SCHEMA public TO PUBLIC");
```

Comment:
Perhaps we should just copy the ACLs from the access node's public schema?

Reply:
> Perhaps we should just copy the ACLs from the access node's public schema?
I don't see how it will help in the general case. Access node and data nodes don't necessarily have the same users with the same permissions, not to mention the fact that access nodes can run an older version of TS (do we support running an older version of PG as well?), which will cause a bit extra problems.

I suggest keeping it as simple as possible.

Reply:
For the record. @pmwkaa responded on Slack that the access node can run an older version of PG than the data node.

Reply:
Well we don't check the PG version I believe only the timescaledb version, do you have a case where this would matter here?

Reply:
@afiskon We don't enforce role consistency across the cluster, but it doesn't mean we don't require it. In fact, multi-node works under the assumption that roles exist across the cluster, at least those that are relevant for multi-node. If they don't exist, things would break down in all kinds of unpredictable ways. 

Thus, it _should_ be an error to bootstrap a database where roles are not consistent across a cluster. It is better to raise an error in that case so that the administrator is aware of the issue.

The issue I have with this statement is that we simply hard-code a GRANT statement on DNs, which might not even exist on the AN. So, when PUBLIC is not granted CREATE on the AN,  it doesn't make much sense to set it on the DN. It should be up to the administrator to set this.

Reply:
@erimatnor Maybe you could give a little more details on what you are suggesting. My understanding is that you suggest the algorithm like:

* List all the user/schema permissions on the access node
* Grant corresponding permissions on the data node

Correct? Should it create the schema if it doesn't exist on the data node? Should it create users as well? How should they be authorized? What if pg_hba.conf differs? Should we somehow check the scenario when there are two users wich accidentally got the same name on the access node and the data node but they serve different purposes (e.g. used by different applications)? Should we deal with something besides users and schemas (tables, views, ....)?

I can't really comment on your proposal since I don't entirely understand it and have a way too many open questions. The most important one is - does the team has resources on this, and did any customer ask us to do anything that complicated? It looks like a rather complicated feature that deserves a separate RFC, review, discussion, and prioritizing. I don't see a reason why it should be part of this patch rather than a separate feature we could add afterward.

I would like to emphasize that we are talking about permission to the **schema**, not to the entire database. My patch doesn't do anything that TimescaleDB doesn't do already on all current PostgreSQL versions. The community just changed the defaults, that's it.

Reply:
@afiskon Forwarding DDL commands (e.g., ALTER/GRANTs and CREATE SCHEMA) to data nodes is something we are actively working on. Discussions, [documentation](https://github.com/timescale/docs/blob/latest/timescaledb/how-to-guides/multinode-timescaledb/multinode-administration.md) and design documents already exist or has happened in the past. But the work is not complete as we are addressing it in steps. 

In the meantime, the basic assumption for multi-node is that, if we don't yet handle something automatically, it is the administrator's responsibility to make sure it is done in a way that is consistent across nodes. This applies to things like role management, settings on the database, etc.

We just recently added the ability to ship schema-related commands to the data nodes (CREATE SCHEMA, ALTER SCHEMA, GRANT/REVOKE), etc., to make it easier for the user to manage schemas across a distributed database.

So, in line with that functionality, schemas and privileges on schemas should be the same across nodes. Now, every change we make need not handle all the cases. For instance, we have not yet implemented the syncing of existing schemas when you bootstrap a new node.

So, for this PR, I think it is enough to at least ship the ACL for the public schema, if we assume that all newly bootstrapped databases on data nodes contain that schema. In most cases that would be true, which I think is good enough. If it is not true, an error will be raised.

It is easy to extend this to additional schemas in the future, but it need not be part of this PR.

So, no, I am not suggesting to snowball this into something big. But neither do I think hard-coding privilege is the right approach.

But, if this is blocking you somehow, we could merge this as an intermediate step to fully deparsing and shipping the schema's ACL if you don't want to do it as part of this PR.

Reply:
Thanks, @erimatnor, I appreciate your detailed reply and patience ❤️. I'm pretty sure I'm missing some context because:

> it is the administrator's responsibility to make sure it is done in a way that is consistent across nodes

... I don't see how the administrator can do this for a new database we just created in create_distributed_hypertable(). So I suggest we make the tests pass and try to do it in a right™ way a bit later :) All in all the patch doesn't make things worse than they are now.

Reply:
I am OK with this temporary approach.